[Yeo Yiheng] iP

yeo-yiheng opened the pull request:
### DukePro
> "A project is complete when it starts working for you, rather than you working for it" - Scott Allen [[source]](https://www.pinterest.com/pin/397864948305207722/)

Ever wanted a personal assistant to keep track of tasks but people are too costly and notebooks are too cumbersome?

DukePro is your **free** and **lightweight** solution which is :
* text-based
* ~~quick~~ _SUPER QUICK_ to access
* simple yet **powerful**

You can own your very own DukePro by:
1. Downloading it from [here.](https://github.com/yeo-yiheng/ip)
2. Run in a terminal.
3. Add your tasks
4. Let it remember and manage your tasks for you 🤩 

Have we mentioned it yet? Or if we did, we shall do so again. It is absolutely **FREE**!

Core Features :
- [X] Managing tasks
- [X] Finding tasks by keyword
- [ ] Reminders (coming soon)


If you're a Java programmer and you'd like to inspect the code, you can inspect the `main` method:
```java
public class Duke {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

yeo-yiheng committed:
Greet, Echo, Exit

yeo-yiheng committed:
Add, List

yeo-yiheng committed:
Mark as Done

yeo-yiheng committed:
ToDos, Events, Deadlines

yeo-yiheng committed:
Automated Text UI Testing

yeo-yiheng committed:
Handle Errors

yeo-yiheng committed:
Delete

yeo-yiheng committed:
Removed comments

yeo-yiheng committed:
Add saving to hard disk compatibility

yeo-yiheng committed:
Add format compatibility for dates and time

yeo-yiheng committed:
Merge branch 'branch-Level-7'

yeo-yiheng committed:
Merge branch-Level-8 to master with branch-Level-7 and resolve conflicts

yeo-yiheng committed:
Ignore tasks.txt

yeo-yiheng committed:
Use More OOP

yeo-yiheng committed:
Organize code into respective packages

yeo-yiheng committed:
Add JUnit tests for storage, parser and subclasses of tasklist

yeo-yiheng committed:
Fix Event functionality to support time range

yeo-yiheng committed:
Add javadocs to the public methods and classes

yeo-yiheng committed:
Follow coding standard guidelines

yeo-yiheng committed:
Add find function to list

yeo-yiheng committed:
Merge branch 'branch-A-JavaDoc'

yeo-yiheng committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/com/duke/task/TaskList.java
#	src/main/java/duke/parser/Parser.java

yeo-yiheng committed:
Merge remote-tracking branch 'origin/add-gradle-support'

yeo-yiheng committed:
Add checkstyle compatibility with gradle

yeo-yiheng committed:
Add gradle dependency for building and running Duke and executing junit tests

Py0000 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,300 @@
+            return true;
+        case "done":
+            String stringIndex = scanner.next();
+            try {
+                int index = Integer.parseInt(stringIndex) - 1;
+                if (task.markDone(index)) {
+                    USER_INTERFACE.taskComplete(TaskList.getTask(index));
+                }
+                break;
+            } catch (NumberFormatException e) {
+                USER_INTERFACE.integerInputWarning();
+                break;
+            }
```
Perhaps you can consider extracting the `break` instead of repeating it in both `try` and `catch` block?

Py0000 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,300 @@
+            TaskList.displayList();
+            break;
+        case "delete":
+            try {
+                String stringIndex2 = scanner.next();
+                int index2 = Integer.parseInt(stringIndex2) - 1;
+                TaskList deletedTask = TaskList.getTask(index2);
+                if (deletedTask == null) {
+                    break;
+                }
+                task.delete(index2);
+                USER_INTERFACE.taskDeleted(deletedTask);
+                break;
+            } catch (NumberFormatException e) {
+                USER_INTERFACE.integerInputWarning();
+                break;
+            }
```
Perhaps you can consider extracting the `break` instead of repeating it in both `try` and `catch` block?

Py0000 commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,77 @@
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Todo;
+import org.junit.jupiter.api.Test;
```
Perhaps 'import org.junit.jupiter.api.Test;` should be separated from the 'duke' import statements? Please do check if this is correct as I am not too sure on this as well.

Py0000 commented on src/test/java/duke/storage/StorageTest.java:
```diff
@@ -0,0 +1,49 @@
+
+import duke.task.TaskList;
+import duke.task.Todo;
+import org.junit.jupiter.api.Test;
```
Similar to the previous suggestion, perhaps import `org.junit.jupiter.api.Test` should be separated from the `duke` import statements?

Py0000 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    /**
+     * This method finds the tasks that matches the input keyword and subsequently
```
Not sure if this is necessary, but perhaps you can consider starting the Javadoc comments with `Finds the tasks...` ?
I have noticed it in some other files as well.
I have also noticed that there is also 2 different style used for the Javadoc comments; one that starts with "This method ..." and other without "This method". Perhaps you can considering to standardize it?

Py0000 reviewed (approved):
I like your program! It was well-coded and easy to read and follow through. 👍 
Minor nitpick: Perhaps you can consider adding some line break in between some statements in some of your codes to further improve readability? 

Other than a few minor nitpicks, it LGTM. 😄 

So sorry that this comment is separated from the rest below 😨

Py0000 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,300 @@
+            return true;
+        case "done":
+            String stringIndex = scanner.next();
+            try {
+                int index = Integer.parseInt(stringIndex) - 1;
+                if (task.markDone(index)) {
+                    USER_INTERFACE.taskComplete(TaskList.getTask(index));
+                }
+                break;
+            } catch (NumberFormatException e) {
+                USER_INTERFACE.integerInputWarning();
+                break;
+            }
```
Perhaps you can consider to extract `break` out instead of writing it in both `try` and `catch` block?

yeo-yiheng replied:
Thank you for this suggestion. I will look into extracting the `break` statement.

Py0000 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,300 @@
+            TaskList.displayList();
+            break;
+        case "delete":
+            try {
+                String stringIndex2 = scanner.next();
+                int index2 = Integer.parseInt(stringIndex2) - 1;
+                TaskList deletedTask = TaskList.getTask(index2);
+                if (deletedTask == null) {
+                    break;
+                }
+                task.delete(index2);
+                USER_INTERFACE.taskDeleted(deletedTask);
+                break;
+            } catch (NumberFormatException e) {
+                USER_INTERFACE.integerInputWarning();
+                break;
+            }
```
Similar to the previous suggestion, perhaps you can consider to extract `break` out instead of writing it in both `try` and catch block?

yeo-yiheng replied:
Thanks! Will do 👍

Py0000 commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,77 @@
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Todo;
+import org.junit.jupiter.api.Test;
```
Perhaps there should be a line break before `import org.junit.jupiter.api.Test;` ? Please do check on this as I am not sure about this as well.

yeo-yiheng replied:
You're right, we should separate related imports. I will change this, thanks for noticing 😄

Py0000 commented on src/test/java/duke/storage/StorageTest.java:
```diff
@@ -0,0 +1,49 @@
+
+import duke.task.TaskList;
+import duke.task.Todo;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
```
Similar to the previous suggestion, perhaps there should be a line break before `import org.junit.jupiter.api.Test;` ? Please do check on this as I am not sure about this as well.

yeo-yiheng replied:
Noted on this! Thank you.

Py0000 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    /**
+     * This method finds the tasks that matches the input keyword and subsequently
```
Not sure if this is required/necessary. Perhaps you can consider to start the Javadoc comments with `Finds the tasks ...` instead of `This method...` ? I have also noticed it in some other Javadoc comments as well.
Also, I have noticed that 2 styles are used for Javadoc comments, namely starting with "This method" and not starting with "This method". Perhaps you can consider to standardize it ?

yeo-yiheng replied:
I agree with you. It should be phrased in a different manner. Thanks for pointing that out and I will make the necessary changes.

yeo-yiheng committed:
Remove binary files and update gitignore to not push class files

Yttruire commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.Scanner;
+
+public class Duke {
+    private static final TaskList TASK = new TaskList();
```
I think that maybe this should also just be named `TASKLIST` or at least `TASKS` (since it is a container for many objects). `TASK` by itself might sound a bit misleading I think.

Yttruire commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,36 @@
+
+        while (SCANNER.hasNext()) {
+            String firstWord = SCANNER.next().toLowerCase(Locale.ROOT);
+            breakWhile = PARSER.firstCommandParser(firstWord, SCANNER);
```
According to coding standards, method names should be verbs, so perhaps this method should be named as `firstCommandParse`?

Yttruire commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,300 @@
+     */
+    public Parser() {}
+
+    private final static UserInterface USER_INTERFACE = new UserInterface();
```
It seems that you have to create more than one UserInterface in your program, which sounds a little weird. Perhaps a suggestion would be for the Parser to have a reference to it rather than create its own one?

Yttruire commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,300 @@
+            }
+            ld = LocalDate.parse(timeFormat);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            USER_INTERFACE.nullFunction();
```
This method might be better as a verb too. At the glance I'm confused what a `nullFunction` is supposed to do

Yttruire commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,103 @@
+
+    private final File file;
+    private final String filePath;
+    private final static UserInterface USER_INTERFACE = new UserInterface();
```
Same as per Parser, a bit strange for every component to have their own user interface

Yttruire reviewed (commented):
Very well done! Can see that you have put a lot of effort into this, with very detailed unit tests, very robust features such as regex as well. Apart from the small points that were brought up, looks great! :D

yeo-yiheng committed:
Add central commands class that oversees all commands, add GUI for improved UX and add javadocs for existing implementation

yeo-yiheng committed:
Add non-null requirements for arguments

yeo-yiheng committed:
Add assertions into program to ensure the state of program is correct during runtime

The program state may be different at times due to the randomized nature of user interaction
with the program. This may cause certain corner cases to slip out of the error handling in place.

Assertions are added at locations which are most likely to encounter errors due to user input.

This is done to optimize the code quality and ensure that assert statements do not clutter
up the code at areas in which there is no inherent need for a state check.

yeo-yiheng committed:
Improved the code quality as per the styling convention

When being reviewed by other people, the code should be of a conventional standard
so as to avoid any erroneous interpretation of code. This commit also ensures that
the code stays up to date with the latest style guide to ensure that it is as readable
as possible.

Inline variables, optimized imports, as well as remove redundant and unused
parameters / variables.

This is done as per the styling guidelines

yeo-yiheng committed:
Fix typo errors in documentation

yeo-yiheng committed:
Merge pull request #1 from yeo-yiheng/branch-A-Assertions

Add assertions into program

yeo-yiheng committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/duke/Duke.java

yeo-yiheng committed:
Add ability to detect duplicate tasks

Whenever a user adds a task, it can get confusing when there are multiple tasks
of the same name. Even if a user has multiple tasks of the same name, it is not
wise to add them as the same task as this is not efficient at all. Thus, it is mandatory
to add tasks of differing names from this commit forth.

Add a way to check the presence of duplicate tasks and disallows users from
creating a task that is already existing.

This is done so that the task list will be easy and intuitive, with clear distinctions
between the tasks

yeo-yiheng committed:
Improve GUI

yeo-yiheng committed:
Add the ui screenshot into docs

yeo-yiheng committed:
Set theme jekyll-theme-modernist

yeo-yiheng committed:
Update README.md

Update README to reflect the updated contents of duke taskbot

yeo-yiheng committed:
Update Readme

yeo-yiheng committed:
Fix file path issue

yeo-yiheng committed:
Remove ui to replace with new ui png

yeo-yiheng committed:
Upload new Ui png
